### PR TITLE
optimize fillcropscpace imagefit

### PR DESCRIPTION
**Pull request by YumengNevix**
1. ensure crop area will not move out image
2. ensure croped image has no white border

### CODE REVIEW

**Comment by ikbendewilliam** [comment]
Works pretty good, one issue I see is that if you zoom out and then in, its not allowed.

File: lib/src/widgets/custom_image_crop_widget.dart [hunk 1]
```diff
@@ -294,7 +295,107 @@ class _CustomImageCropState extends State<CustomImageCrop>
...
+  }
+
+   // just support 'pi/2 * n' angle for now
+  bool _isAngleSupported() {
+    final angle = data.angle;
+    return angle % (pi/2) == 0;
+  }
+
+  void _correctTransition(CropImageData transition, VoidCallback callback) {
+    if (_imageAsUIImage == null ||
```
Comment (ikbendewilliam) [inline lib/src/widgets/custom_image_crop_widget.dart:330]: Also add an option in the widget `forceInsideCropArea` and disable it by default so it's not a breaking change + there is no unexpected behaviour.

Comment (ikbendewilliam) [reply lib/src/widgets/custom_image_crop_widget.dart:330]: Also add this argument/property to the readme

**Review by ikbendewilliam — CHANGES_REQUESTED** [review]

**Comment by YumengNevix** [comment]
> Works pretty good, one issue I see is that if you zoom out and then in, its not allowed.

Pre implementation only support angle which is integer multiple of pi/2,  latest commit supports any angle, so this issue should be fixed.

**Comment by ikbendewilliam** [comment]
Some great work already!
The angle wasn't the issue, previously if I zoomed out using the buttons everything was good, then zooming in with the buttons would result in a zoom level of `0,9975` which didn't work (1.33 x 0.75), I guess that was on the numbers I used, if we used 1.34 it would have worked. 
That being said, the zoom buttons now don't work at all now when this is enabled (unless I first zoom out with gestures), which is a bigger issue. 
Also if I turn left once, then turn right once (using the buttons), the image is zoomed out.

**Comment by YumengNevix** [comment]
> Some great work already! The angle wasn't the issue, previously if I zoomed out using the buttons everything was good, then zooming in with the buttons would result in a zoom level of `0,9975` which didn't work (1.33 x 0.75), I guess that was on the numbers I used, if we used 1.34 it would have worked. That being said, the zoom buttons now don't work at all now when this is enabled (unless I first zoom out with gestures), which is a bigger issue. Also if I turn left once, then turn right once (using the buttons), the image is zoomed out.

1.  The zoom out bug is caused by _getDistanceBetweenPointAndLine, it may result in NaN，so I prevent the scale op.
2. The zoom in bug is caused by _calculateScaleAfterRotate, when data.scale is less than 1.0, startScale is 1.33, endScale is 1.4, binary search range should be [data.scale, endScale] rather than [startScale, endScale], after modify implementation, it will result in 1.0 rather than 1.33.
3. When user rotate, scale will change automatically to ensure fully cover path, it's by design, not bug.

**Comment by ikbendewilliam** [comment]
Works great! Thank you!

**Comment by ikbendewilliam** [comment]
released in 0.0.13

**Pull request merged**